### ISSUE TITLE
Wrong results given by triton bwd

### ISSUE THREAD

**Issue by LouChao98** [lucidrains/CoLT5-attention#8]
https://github.com/lucidrains/CoLT5-attention/blob/860081fcaabfc0125e8530507f4bf8d802596281/colt5_attention/triton_coor_descent.py#L291

here should be `last_da = tl.sum(ds, axis = 0)`

test script: 
```python

import torch
from torch.autograd import grad
from colt5_attention import coor_descent
from colt5_attention.triton_coor_descent import triton_coor_descent

s = torch.randn(512).cuda().requires_grad_()

result1 = coor_descent(s, n_iters=50, k=64, eps=1.0)
result2 = triton_coor_descent(s, n_iters=50, k=64, eps=1.0, checkpoint_segments=1)

g1 = grad(result1[:32].sum(), s)[0]
g2 = grad(result2[:32].sum(), s)[0]
print((g1 - g2).norm())
```
---

A suggestion. I recommand to add a type conversion here https://github.com/lucidrains/CoLT5-attention/blob/860081fcaabfc0125e8530507f4bf8d802596281/colt5_attention/transformer_block.py#L436-L438 `s.to(torch.float32)`. Then the code works in half-precision training.

**Comment by lucidrains** [comment]
@LouChao98 hello! yes you are right that it should be done in float32. disabled the autocast through a decorator in 0.10.16. do let me know if that is enough

so the `last_da` is to account for the gradients due to the final summation of `a` to the final output [here](https://github.com/lucidrains/CoLT5-attention/blob/main/colt5_attention/triton_coor_descent.py#L390). i'm seeing errors on the order of `~3e-7` in the script you provided. what are you seeing?

**Comment by LouChao98** [comment]
I use the latest triton-nightly.

**Comment by lucidrains** [comment]
@LouChao98 how did you install triton? i did `pip install triton --pre -U`. it shows that i have triton 2.1 installed

**Comment by lucidrains** [comment]
@LouChao98 yes, i was actually just looking into that... it was prompted by your initial issue

**Comment by lucidrains** [comment]
@LouChao98 ok, let me try

i want this repository to be perfect, as coordinate descent is very promising

**Comment by lucidrains** [comment]
ah, the latest version doesn't even run for me

```bash
LLVM ERROR: Cannot select: intrinsic %llvm.nvvm.shfl.sync.bfly.i32
```

**Comment by lucidrains** [comment]
@LouChao98 you don't happen to have a working machine you can lend me to debug this in the next few hours? lol

**Comment by LouChao98** [comment]
I use your way to install triton, and I get `~1e-8` error. 
I test two machines using my way to install triton and get similar large error.

I believe this is a triton issue.

**Comment by LouChao98** [comment]
> @LouChao98 you don't happen to have a working machine you can lend me to debug this in the next few hours? lol

Sorry, my machine is not accessable publicly. It is a little hard to share with you.

**Comment by lucidrains** [comment]
@LouChao98 regardless, i think you are correct that `last_da` should only be applied on the last backwards segment (i actually wouldn't have thought about this if you haven't raised the issue so thanks!)

**Comment by lucidrains** [comment]
@LouChao98 cool! i didn't actually see much of anything with mixture of attention, but ymmv

gl!
